RapidReact/src/main/java/frc/robot/subsystem/ClimberSubsystem.java

```diff
@@ -1,112 +1,311 @@
+        new DoubleSolenoid(PneumaticsModuleType.REVPH, config.elevatorSolenoid_ID1, config.elevatorSolenoid_ID2);
     }
+
+    currentClimbState = ClimbState.Idle;
+  }
+
+  boolean isClimberAtSetpoint()
+  {
+    climbErrThreshold1 = 10;
+    climbLoopsToSettle1 = 10;
+    withinThresholdLoops1 = 0;
+    boolean climber1AtSetpoint = false;
+    if (climber1.getClosedLoopError() < +climbErrThreshold1 && climber1.getClosedLoopError() > -climbErrThreshold1) {
+      withinThresholdLoops1++;
+
+      if (withinThresholdLoops1 > climbLoopsToSettle1)
+        climber1AtSetpoint = true;
+    } else {
+      withinThresholdLoops1 = 0;
+    }
+    
+    climbErrThreshold2 = 10;
+    climbLoopsToSettle2 = 10;
+    withinThresholdLoops2 = 0;
+    boolean climber2AtSetpoint = false;
+    if (climber2.getClosedLoopError() < +climbErrThreshold2 && climber2.getClosedLoopError() > -climbErrThreshold2) {
+      withinThresholdLoops2++;
+
+      if (withinThresholdLoops2 > climbLoopsToSettle2)
+        climber2AtSetpoint = true;
+    } else {
+      withinThresholdLoops2 = 0;
+    }
+
+    return climber1AtSetpoint && climber2AtSetpoint;
+  }
+
+  private void idleInit()
+  {
+    setElevatorTilted(false);
+    autoRetractFull();
+  }
+
+  private void extendPartialInit()
+  {
+    setElevatorTilted(false);
+    autoExtendPartial();
+  }
+
+  private void extendFullInit()
+  {
+    setElevatorTilted(true);
+    autoExtendFull();
+  }
+
+  private void retractInit()
+  {
+    setElevatorTilted(false);
+    autoRetractFull();
   }
 
   @Override
-  public void periodic() {}
+  public void periodic() {
+    if (!autoClimb)
+      return;
+    
+    ClimbState nextState = ClimbState.Idle;
+
+    switch (currentClimbState)
+    {
+      case Idle:
+        if (autoClimbPressed)
+        {
+          nextState = ClimbState.ExtendPartial;
+          extendPartialInit();
+        }
+        else
+        {
+          nextState = ClimbState.Idle;
+        }
+        break;
+    
+      case ExtendPartial:
+        if (isClimberAtSetpoint())
+        {
+          nextState = ClimbState.ExtendFull;
+          extendFullInit();
+        }
+        else
+        {
+          nextState = ClimbState.ExtendPartial;
+        }
+        break;
+      
+      case ExtendFull:
+        if (isClimberAtSetpoint())
+        {
+          nextState = ClimbState.Retract;
+          retractInit();
+        }
+        else
+        {
+          nextState = ClimbState.ExtendFull;
+        }
+        break;
+      
+      case Retract:
+        if (isClimberAtSetpoint())
+        {
+          nextState = ClimbState.Idle;
+          idleInit();
+        }
+        else
+        {
+          nextState = ClimbState.Retract;      
+        }
+        break;
+    }
+
+    currentClimbState = nextState;
+  }
+
 
   @Override
   public void disable() {
-    climber.set(0);
+    climber1.set(0);
+    climber2.set(0);
   }
 
   public void toggleClimberEnabled() { // uses 2 PS button
```

Comment:
Don't use "toggle" since that has additional hidden state that the driver or the code has to keep track of. Use "setClimberEnabled(bool enabled)". And give the driver different enable and disable buttons.

Reply:
Our climber is "enabled" when both the driver and operator press the PS4 buttons at the same time. It's an end game thing. I suppose we could add an extra "disable" button combination so they don't accidentally toggle it back, but this "untoggle" is mostly for testing.